code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/HostAffinityConstraint.java

```diff
@@ -33,16 +34,26 @@ public boolean matches(AllocationAttempt attempt,
         Set<Long> hostIds = candidate.getHosts();
         if (op == AffinityOps.SOFT_EQ || op == AffinityOps.EQ) {
             for (Long hostId : hostIds) {
-                Map<String, String> labelsForHost = allocatorDao.getLabelsForHost(hostId);
-                if (!labelValue.equals(labelsForHost.get(labelKey))) {
+                Map<String, String[]> labelsForHost = allocatorDao.getLabelsForHost(hostId);
+                if (labelsForHost.get(labelKey) == null) { // key doesn't exist
+                    return false;
+                }
+                String value = labelsForHost.get(labelKey)[0];
+                String hostLabelMapState = labelsForHost.get(labelKey)[1];
+                if (!labelValue.equals(value)) { // value doesn't match
+                    return false;
+                }
+                if (CommonStatesConstants.REMOVING.equals(hostLabelMapState)) { // value matches but it's currently getting removed
                     return false;
                 }
             }
             return true;
         } else {
             for (Long hostId : hostIds) {
-                Map<String, String> labelsForHost = allocatorDao.getLabelsForHost(hostId);
-                if (labelValue.equals(labelsForHost.get(labelKey))) {
+                Map<String, String[]> labelsForHost = allocatorDao.getLabelsForHost(hostId);
+                if (labelsForHost.get(labelKey) != null 
```

Comment:
I just checked and I think we should be ok.  I did notice that labelKey could potentially be null, but it looks like that should be ok.